src/api/general-services/pipeline-manage/index.js

```diff
@@ -0,0 +1,176 @@
+    }
+    default: {
+      throw new Error(`Invalid state type specified: ${stepType}`);
+    }
+  }
+  /* eslint-enable global-require */
+};
+
+const createNewStateMachine = async (context, stateMachine) => {
+  const { clusterEnv, sandboxId } = config;
+  const { experimentId, roleArn, accountId } = context;
+
+  const stepFunctions = new AWS.StepFunctions({
+    region: config.awsRegion,
+  });
+
+  const pipelineHash = crypto
+    .createHash('sha1')
+    .update(`${experimentId}-${sandboxId}`)
+    .digest('hex');
+
+  const params = {
+    name: `biomage-pipeline-${pipelineHash}`,
+    roleArn,
+    definition: JSON.stringify(stateMachine),
+    loggingConfiguration: { level: 'OFF' },
+    tags: [
+      { key: 'experimentId', value: experimentId },
+      { key: 'clusterEnv', value: clusterEnv },
+      { key: 'sandboxId', value: sandboxId },
+    ],
+    type: 'STANDARD',
+  };
+
+  let stateMachineArn = null;
+
+  try {
+    const response = await stepFunctions.createStateMachine(params).promise();
+    stateMachineArn = response.stateMachineArn;
+  } catch (e) {
+    if (e.code !== 'StateMachineAlreadyExists') {
+      throw e;
+    }
+
+    stateMachineArn = `arn:aws:states:${config.awsRegion}:${accountId}:stateMachine:${params.name}`;
+
+    await stepFunctions.updateStateMachine(
+      { stateMachineArn, definition: params.definition, roleArn },
+    ).promise();
+  }
+
+  return stateMachineArn;
+};
+
+const executeStateMachine = async (stateMachineArn) => {
+  const stepFunctions = new AWS.StepFunctions({
+    region: config.awsRegion,
+  });
+  const { trace_id: traceId } = AWSXRay.getSegment() || {};
+
+  const { executionArn } = await stepFunctions.startExecution({
+    stateMachineArn,
+    input: '{}',
+    traceHeader: traceId,
+  }).promise();
+
+  return executionArn;
+};
+
+const createPipeline = async (experimentId) => {
+  const accountId = await config.awsAccountIdPromise();
+  const roleArn = `arn:aws:iam::${accountId}:role/state-machine-role-${config.clusterEnv}`;
+
+  const context = {
+    experimentId,
+    accountId,
+    roleArn,
+    experiemntId: experimentId,
+    pipelineImages: await getPipelineImages(),
+    clusterInfo: await getClusterInfo(),
+  };
+
+  const skeleton = {
+    Comment: 'N/A',
+    StartAt: 'DeleteCompletedPipelineWorker',
+    States: {
+      DeleteCompletedPipelineWorker: {
+        XStepType: 'delete-completed-jobs',
+        Next: 'LaunchNewPipelineWorker',
+      },
+      LaunchNewPipelineWorker: {
+        XStepType: 'create-new-job-if-not-exist',
+        Next: 'Wait',
+      },
+      Wait: {
+        Type: 'Wait',
+        Seconds: 5,
+        End: true,
+      },
+    },
+  };
+
+  logger.log('Constructing pipeline from step skeleton...');
+  const stateMachine = _.cloneDeepWith(skeleton, (o) => {
+    if (_.isObject(o) && o.XStepType) {
+      return { ...constructPipelineStep(context, o), XStepType: undefined };
+    }
+    return undefined;
+  });
```

Comment:
This is basically a reducer. It goes through the skeleton and depending on the value of `XStepType` it will call into a different reducer to change the object slightly.